File: cortex/core/types/doctor.py
"""
Doctor module types for diagnostic results.

These types are used by the doctor module to return structured
diagnosis information about metrics, variants, and other entities.
"""

from typing import List, Optional, Union

from cortex.core.semantics.metrics.metric import SemanticMetric
from cortex.core.semantics.metrics.variant import SemanticMetricVariant
from cortex.core.types.telescope import TSModel


class Suggestion(TSModel):
    """
    A concrete fix suggestion for an identified issue.

    Each suggestion includes a description of why the error occurred
    and a corrected entity JSON that resolves the issue.
    """

    description: str
    # Put SemanticMetricVariant FIRST - Pydantic v2 smart mode will use runtime type
    # SemanticMetricVariant has unique required field 'source' for better matching
    fixed: Union[SemanticMetricVariant, SemanticMetric]


class Diagnosis(TSModel):
    """
    Full diagnosis result containing an explanation and fix suggestions.

    The explanation covers ALL detected issues (both fixable and non-fixable).
    Suggestions only contain issues where a concrete fix could be generated.
    """

    explanation: str
    suggestions: List[Suggestion]


class DiagnosisResult(TSModel):
    """
    Top-level result returned by the doctor.

    When healthy is True, diagnosis is None (no issues found).
    When healthy is False, diagnosis contains the full analysis.
    """

    healthy: bool
    diagnosis: Optional[Diagnosis] = None

File: frontend/cortex/app/types/doctor.ts
/**
 * Type definitions for the Doctor (Diagnose) API.
 *
 * Mirrors backend types from cortex/core/types/doctor.py and
 * cortex/sdk/schemas/requests/doctor.py / responses/doctor.py
 */

// ============================================================================
// Response Types
// ============================================================================

export interface Suggestion {
  /** Why the error occurred */
  description: string
  /** Corrected entity JSON that resolves the issue */
  fixed: Record<string, any>
}

export interface Diagnosis {
  /** Human-readable summary of ALL detected issues */
  explanation: string
  /** Concrete fix suggestions (only for issues where a fix could be generated) */
  suggestions: Suggestion[]
}

export interface DiagnoseResponse {
  /** Whether the entity is healthy (no issues detected) */
  healthy: boolean
  /** Full diagnosis result (null when healthy is true) */
  diagnosis: Diagnosis | null
}

// ============================================================================
// Request Types
// ============================================================================

export interface MetricDiagnoseRequest {
  metric_id?: string
  metric?: Record<string, any>
  environment_id: string
}

export interface VariantDiagnoseRequest {
  variant_id?: string
  variant?: Record<string, any>
  environment_id: string
}
